trieste/acquisition/function/utils.py

```diff
@@ -0,0 +1,252 @@
+        :param mean: Tensor of shape (B, Q), batch of means.
+        :param covariance: Tensor of shape (B, Q, Q), batch of covariances.
+        :param jitter: float, jitter to use in the Cholesky factorisation.
+        :returns mvn_cdf: Tensor of shape (B,), CDF values.
+        """
+
+        # Check shapes of input tensors
+        tf.debugging.assert_shapes(
+            [
+                (x, ("B", "Q")),
+                (mean, ("B", "Q")),
+                (cov, ("B", "Q", "Q")),
+            ]
+        )
+
+        # Identify data type to use for all calculations
+        dtype = mean.dtype
+        B, Q = mean.shape
+
+        # Compute Cholesky factors
+        jitter = jitter * tf.eye(Q, dtype=dtype)[None, :, :]
+        C = tf.linalg.cholesky(cov + jitter)  # (B, Q, Q)
+
+        # Rename samples and limits for brevity
+        w = samples  # (S, Q)
+        b = x - mean  # (B, Q)
+
+        # Initialise transformation variables
+        e = tf.zeros(shape=(B, S, Q), dtype=dtype)
+        f = tf.zeros(shape=(B, S, Q), dtype=dtype)
+        y = tf.zeros(shape=(B, S, Q), dtype=dtype)
+
+        # Initialise standard normal for computing CDFs
+        Phi, iPhi = standard_normal_cdf_and_inverse_cdf(dtype=dtype)
+
+        # Get update indices for convenience later
+        idx = get_update_indices(B=B, S=S, Q=Q, q=0)
+
+        # Slice out common tensors
+        b0 = b[:, None, 0]
+        C0 = C[:, None, 0, 0] + 1e-12
+
+        # Compute transformation variables at the first step
+        e_update = tf.tile(Phi(b0 / C0), (1, S))  # (B, S)
+        e = tf.tensor_scatter_nd_add(e, idx, e_update)
+        f = tf.tensor_scatter_nd_add(f, idx, e_update)
+
+        for i in tf.range(1, Q):
+
+            # Update y tensor
+            y_update = iPhi(1e-6 + (1 - 2e-6) * w[None, :, i - 1] * e[:, :, i - 1])
+            y = tf.tensor_scatter_nd_add(y, idx, y_update)
+
+            # Slice out common tensors
+            bi = b[:, None, i]
+            Ci_ = C[:, None, i, :i]
+            Cii = C[:, None, i, i] + 1e-12
+            yi = y[:, :, :i]
+
+            # Compute indices to update d, e and f tensors
+            idx = get_update_indices(B=B, S=S, Q=Q, q=i)
+
+            # Update e tensor
+            e_update = Phi((bi - tf.reduce_sum(Ci_ * yi, axis=-1)) / Cii)
+            e = tf.tensor_scatter_nd_add(e, idx, e_update)
+
+            # Update f tensor
+            f_update = e[:, :, i] * f[:, :, i - 1]
+            f = tf.tensor_scatter_nd_add(f, idx, f_update)
+
+        mvn_cdf = tf.reduce_mean(f[:, :, -1], axis=-1)
+
+        return mvn_cdf
+
+    return mvn_cdf
+
+
+# =============================================================================
+# Multivariate Normal Expected Improvement using naive Monte Carlo estimation
+# =============================================================================
+
+
+@tf.function
+def gaussian_monte_carlo_expected_improvement(
+    mean: TensorType,
+    covariance: TensorType,
```

Comment:
why is this included? This is just a bad version of BatchMonteCarloEI?

Reply:
I included this for doing some preliminary testing, but it's not needed so I'll remove it. Thanks for reminding me!